# Part 1. Equation of a Slime

How many late days are you using for this assignment?

In [28]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import ShuffleSplit
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import make_pipeline

## 1. Loading the dataset

In [3]:
# Using pandas load the dataset
data = pd.read_csv('science_data_large.csv')
# Output the first 15 rows of the data
print("First 15 Rows of the Dataset:")
print(data.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print("\nSummary of the Dataset:")
print(data.info())
print(data.describe())


First 15 Rows of the Dataset:
    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

Summary of the Dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL    

## 2. Splitting the dataset

In [4]:
# Take the pandas dataset and split it into our features (X) and label (y)

X = data[['Temperature °C', 'Mols KCL']]  # Adjust the feature columns as necessary
y = data['Size nm^3']  # Adjust the label column name accordingly

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.10, random_state=42)

# For grading consistency use random_state=42 

print(f"Training set size: {X_train.shape[0]} samples")
print(f"Test set size: {X_test.shape[0]} samples")

Training set size: 900 samples
Test set size: 100 samples


## 3. Perform a Linear Regression

In [6]:
# Use sklearn to train a model on the training set

model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model

sample_datapoint = [[25, 0.5]]  # 25 °C, 0.5 Mols KCL
predicted_size = model.predict(sample_datapoint)

print(f"Predicted Size for sample datapoint (Temperature: 25°C, Mols KCL: 0.5): {predicted_size[0]} nm^3")

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means

score = model.score(X_test, y_test)
print(f"Model Score: {score}")

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX

coefficients = model.coef_
intercept = model.intercept_

# For example, using coefficients list: [coef_temp, coef_kcl]
print(f"Coefficients: {coefficients}")
print(f"Intercept: {intercept}")

# LaTeX equation for the model
equation = f"h(x) = {intercept:.5f} + ({coefficients[0]:.5f} * Temperature) + ({coefficients[1]:.5f} * Mols KCL)"
print(f"Equation: {equation}")

Predicted Size for sample datapoint (Temperature: 25°C, Mols KCL: 0.5): -387221.4717158666 nm^3
Model Score: 0.8552472077276095
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.4795834075
Equation: h(x) = -409391.47958 + (866.14641 * Temperature) + (1032.69507 * Mols KCL)


/Users/<redacted>/Library/Python/3.11/lib/python/site-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


## **Linear Equation for Slime Growth**  

From our trained model, the equation predicting slime size based on temperature and KCL concentration is:  

\[
h(x) = -409391.47958 + (866.14641 *Temperature) + (1032.69507 * Mols KCL)
\]

##### **The General Equation**: h(x)= β0 + β1(Temperature ∘C) + β2(Mols KCL)

This equation suggests that for each additional degree Celsius, the predicted slime size increases by **866.15 nm³**, and for each additional **1 mol** of KCL, the slime size increases by **1032.70 nm³**.  

---

## **Model Score and Explanation**  

The model achieved a score of **0.855**, which refers to the **coefficient of determination** (\( R^2 \) score). This metric measures how well the model's predictions match the actual data.  

- An \( R^2 \) score of **1.0** indicates a perfect fit, meaning the model explains **100%** of the variance in slime growth.  
- An \( R^2 \) score of **0** means the model performs no better than predicting the average slime size.  
- Our score of **0.855** means that about **85.5% of the variance** in slime size is explained by the model. This suggests a **strong fit**, but there is still some room for improvement.  

## 4. Use Cross Validation

In [18]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42

cv = ShuffleSplit(n_splits=5, random_state=42)
scores = cross_val_score(model, X, y, cv=cv)

# Report on their finding and their significance
print(f"Cross Validation Scores: {scores}")
print(f"Mean Score: {scores.mean():.5f}")
print(f"Standard Deviation: {scores.std():.5f}")

Cross Validation Scores: [0.85524721 0.85830729 0.8654508  0.83855498 0.83002469]
Mean Score: 0.84952
Standard Deviation: 0.01315


### Findings from Cross-Validation

The following results were obtained using 5-fold cross-validation with the linear regression model on the dataset:

- **Cross Validation Scores**: 
  - [0.85524721 0.85830729 0.8654508  0.83855498 0.83002469]

- **Mean Score**: 
  - **0.84952**

- **Standard Deviation**: 
  - **0.01315**

#### Interpretation of Findings

1. **Mean Score**: 
   - The mean score of **0.84952** indicates that, on average, the model explains approximately **84.95%** of the variance in the target variable, which in this context represents the size of the slime. This suggests that the linear regression model demonstrates solid predictive capability when generalized across different subsets of the data through cross-validation.

2. **Cross-Validation Scores**: 
   - The individual cross-validation scores show a range from **0.8300** to **0.8655**, which demonstrates that the model's performance is relatively consistent across different data shuffles. The highest score (**0.8655**) indicates a strong fit for some subsets, while the lowest score (**0.8300**) suggests some variability in predictions.

3. **Standard Deviation**:
   - The standard deviation of **0.01315** is low. This indicates that the model’s performance is stable and does not vary significantly between different training and testing data splits. A low standard deviation suggests that the model is reliable, reinforcing the results obtained from the mean score.

### Conclusion

In summary, the results from the cross-validation indicate that the linear regression model is performing well in predicting the size of slime based on the features of the dataset. The model's consistency and reasonable predictive power suggest that it appropriately captures the relationship between the independent variables (temperature and concentration of KCl) and the dependent variable (size of the slime).


## 5. Using Polynomial Regression

In [ ]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score, KFold


# Step 1: Augment the original feature set using degree 2 polynomial features
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)

# Step 2: Create a Linear Regression model
model = LinearRegression()

# Step 3: Perform 5-fold cross validation with shuffling for reproducibility
kf = KFold(n_splits=5, shuffle=True, random_state=42)
scores_poly = cross_val_score(model, X_poly, y, cv=kf)

# Step 4: Report the mean and standard deviation of the cross-validation scores
mean_score_poly = scores_poly.mean()
std_dev_score_poly = scores_poly.std()
print(f"Cross-Validation Scores (Polynomial Features): {scores_poly}")
print(f"Mean Score (Polynomial Features): {mean_score_poly:.4f}")
print(f"Standard Deviation of Scores (Polynomial Features): {std_dev_score_poly:.4f}")

# Step 5: Fit the model to the full dataset of polynomial features
model.fit(X_poly, y)

# Step 6: Extract coefficients and intercept from the trained model
intercept_poly = model.intercept_
coefficients_poly = model.coef_
print(f"Intercept: {intercept_poly}")
print(f"Coefficients: {coefficients_poly}")

# Generate the equation in LaTeX
# If X is a DataFrame, we can extract the original feature names
feature_names = poly.get_feature_names_out(X.columns)
equation = f"h(x) = {intercept_poly:.5f}"
for name, coef in zip(feature_names, coefficients_poly):
    equation += f" + ({coef:.5f} \\times {name})"

print("LaTeX Equation:")
print(equation)


Cross-Validation Scores (Polynomial Features): [0.69121588 0.69407705 0.73980045 0.61215057 0.4412892 ]
Mean Score (Polynomial Features): 0.6357
Standard Deviation of Scores (Polynomial Features): 0.1055
Intercept: 1.3304255415455621
Coefficients: [-1.50569525e-15  3.46146199e-01 -7.25384665e-01  3.76179350e-01
 -1.31721534e-01  1.31721534e-01 -2.27440660e+00 -8.79905473e-03
  6.81978938e-01  9.74543422e-02  2.86871570e-02  3.17459042e-01
 -1.52408729e+00  2.38056811e-01  1.66585764e+00 -5.38257980e-01
 -1.87126685e-01  5.23232232e-01 -7.02349978e-01 -3.87908185e-01
  7.64087536e-01 -1.90013696e+00 -1.31721534e-01  0.00000000e+00
  1.34221685e+00  1.31721534e-01 -3.61662345e+00  5.56215209e+00]
LaTeX Equation:
h(x) = 1.33043 + (-0.00000 \times 1) + (0.34615 \times age) + (-0.72538 \times bp) + (0.37618 \times wbcc) + (-0.13172 \times appet_poor) + (0.13172 \times appet_good) + (-2.27441 \times rbcc) + (-0.00880 \times age^2) + (0.68198 \times age bp) + (0.09745 \times age wbcc) + (0.02

## Polynomial Regression (Degree 2) for Slime Growth

We augmented the original feature set using the `PolynomialFeatures` library with a degree of 2. This process transforms the original features into an expanded set that includes:
- The original features (e.g., `age`, `bp`, `wbcc`, `appet_poor`, `appet_good`, `rbcc`)
- Squared terms (e.g., `age^2`, `bp^2`, etc.)
- Interaction terms (e.g., `age bp`, `age wbcc`, etc.)

### Steps Followed:

1. **Feature Augmentation:**
   - The original features in **X** were transformed into polynomial features of degree 2 using `PolynomialFeatures(degree=2)`.

2. **Model Creation:**
   - We created a `LinearRegression` model to fit these augmented features.

3. **Cross-Validation:**
   - We performed 5-fold cross-validation (with shuffling and `random_state=42` for reproducibility) using the transformed features.
   - **Cross-Validation Scores:**  
     The individual fold scores were:  
     ` [0.6912, 0.6941, 0.7398, 0.6122, 0.4413]`
   - **Mean Score:**  
     The average R² score was **0.6357**.
   - **Standard Deviation:**  
     The standard deviation of the scores was **0.1055**, indicating moderate variability across folds.

4. **Model Fitting:**
   - The model was then fitted on the full augmented dataset to obtain the final coefficients and intercept.

5. **Resulting Equation:**
   - The extracted intercept and coefficients were used to generate the following polynomial regression equation:
   
   \[
   \begin{aligned}
   h(x) =\,& 1.33043 + (-0.00000 \times 1) + (0.34615 \times \text{age}) + (-0.72538 \times \text{bp}) \\
   & + (0.37618 \times \text{wbcc}) + (-0.13172 \times \text{appet\_poor}) + (0.13172 \times \text{appet\_good}) \\
   & + (-2.27441 \times \text{rbcc}) + (-0.00880 \times \text{age}^2) + (0.68198 \times \text{age bp}) \\
   & + (0.09745 \times \text{age wbcc}) + (0.02869 \times \text{age appet\_poor}) + (0.31746 \times \text{age appet\_good}) \\
   & + (-1.52409 \times \text{age rbcc}) + (0.23806 \times \text{bp}^2) + (1.66586 \times \text{bp wbcc}) \\
   & + (-0.53826 \times \text{bp appet\_poor}) + (-0.18713 \times \text{bp appet\_good}) + (0.52323 \times \text{bp rbcc}) \\
   & + (-0.70235 \times \text{wbcc}^2) + (-0.38791 \times \text{wbcc appet\_poor}) + (0.76409 \times \text{wbcc appet\_good}) \\
   & + (-1.90014 \times \text{wbcc rbcc}) + (-0.13172 \times \text{appet\_poor}^2) + (0.00000 \times \text{appet\_poor appet\_good}) \\
   & + (1.34222 \times \text{appet\_poor rbcc}) + (0.13172 \times \text{appet\_good}^2) \\
   & + (-3.61662 \times \text{appet\_good rbcc}) + (5.56215 \times \text{rbcc}^2)
   \end{aligned}
   \]
   
   *Note: The actual printed equation in the console includes the precise values and uses the feature names extracted from the DataFrame.*

### Interpretation of the Scores:

- The **mean R² score** of **0.6357** suggests that the polynomial regression model explains approximately **63.57%** of the variance in the target variable on average across the folds.
- The **standard deviation** of **0.1055** indicates that while there is some variability in performance across different data splits, the model's performance is relatively consistent.
- Overall, these metrics imply that the degree-2 polynomial features have provided a decent but not perfect explanation of the data. This may indicate that the underlying relationship could be more complex, or that other features or higher-degree terms might be needed to improve the model.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [49]:
# Load the dataset. Then train and evaluate the classification models.
data = pd.read_csv('ckd_feature_subset.csv')

print(data.head(15))
print("\nSummary of the Dataset:")
print(data.info())

         age        bp      wbcc  appet_poor  appet_good      rbcc  Target_ckd
0   0.688312  0.333333  0.000000           1           0  0.000000           1
1   0.545455  0.333333  0.128319           1           0  0.305085           1
2   0.714286  0.500000  0.238938           1           0  0.186441           1
3   0.688312  0.333333  0.283186           0           1  0.338983           1
4   0.441558  0.333333  0.221239           1           0  0.220339           1
5   0.740260  0.833333  0.265487           0           1  0.355932           1
6   0.870130  0.333333  0.668142           0           1  0.237288           1
7   0.870130  0.500000  0.150442           0           1  0.372881           1
8   0.194805  0.666667  0.380531           0           1  0.305085           1
9   0.753247  0.833333  0.163717           0           1  0.220339           1
10  0.701299  0.166667  0.504425           1           0  0.152542           1
11  0.610390  0.666667  0.367257           1        

In [ ]:
# Define features (X) and target labels (y)
X = data.drop('Target_ckd', axis=1)  # Features
y = data['Target_ckd']                 # Target variable

# Import train_test_split
from sklearn.model_selection import train_test_split

# Split the dataset into training and test sets (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.10, random_state=42)

# Confirm the sizes of train and test sets
print(f"Training set size: {X_train.shape[0]} samples")
print(f"Test set size: {X_test.shape[0]} samples")

Training set size: 137 samples
Test set size: 16 samples


## Results and Conclusion for Classification Experiments

In [48]:
# Import necessary libraries for classification
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import cross_val_score

# Initialize models
models = {
    "Logistic Regression": LogisticRegression(max_iter=1000),
    "Support Vector Machine": SVC(),
    "K-Nearest Neighbors": KNeighborsClassifier(),
    "Neural Network": MLPClassifier(max_iter=1000, random_state=42)
}

# Evaluate models using 5-fold cross-validation
results = {}
for name, model in models.items():
    scores = cross_val_score(model, X_train, y_train, cv=5)
    results[name] = {
        "mean_score": scores.mean(),
        "std_dev": scores.std()
    }

# Convert results to DataFrame for better visualization
results_df = pd.DataFrame(results).T
results_df

,mean_score,std_dev
Logistic Regression,0.867725,0.061225
Support Vector Machine,0.918783,0.071983
K-Nearest Neighbors,0.911640,0.072957
Neural Network,0.948413,0.037878


### Findings:

- The **Neural Network** model achieved the highest mean score of **0.948**, indicating it is the most effective classifier for this dataset among those evaluated. The low standard deviation of **0.038** suggests that the model is stable across different folds.
- The **Support Vector Machine (SVM)** model closely follows with a mean score of **0.919**. This model also performed very well, suggesting it can effectively delineate the decision boundaries between the classes. However, it was slightly less effective than the neural network.
- The **K-Nearest Neighbors** model had a mean score of **0.9116**. While still a robust classifier, it was outperformed by both the SVM and the Neural Network. The relatively high variance (standard deviation of **0.07296**) suggests that KNN may be sensitive to the specific partition of the data used in cross-validation.
- The **Logistic Regression** model achieved the lowest mean score of **0.8677** among the models evaluated. Despite being a simpler model, it still produced reasonable results. However, its performance indicates that it may not capture the complexities of the dataset as effectively as the other three models. The standard deviation of 0.0612 suggests a stable performance across the folds, though lower overall accuracy.


Overall, the Neural Network is concluded to be the most suitable model for this classification task based on the feature set available.

## Results and Conclusion for Neural Network Experiments

In [57]:
# Experimenting with different settings for the Neural Network
nn_configs = [
    {'hidden_layer_sizes': (10,), 'activation': 'logistic'},
    {'hidden_layer_sizes': (10,), 'activation': 'relu'},
    {'hidden_layer_sizes': (20,), 'activation': 'tanh'},
]

# Reporting findings for different configurations
nn_results = {}
for config in nn_configs:
    nn_model = MLPClassifier(hidden_layer_sizes=config['hidden_layer_sizes'], activation=config['activation'], max_iter=2000, random_state=42)
    scores = cross_val_score(nn_model, X_train, y_train, cv=5)
    nn_results[f"Layers: {config['hidden_layer_sizes']}, Activation: {config['activation']}"] = {
        "mean_score": scores.mean(),
        "std_dev": scores.std()
    }

nn_results_df = pd.DataFrame(nn_results).T
nn_results_df

,mean_score,std_dev
"Layers: (10,), Activation: logistic",0.911905,0.055830
"Layers: (10,), Activation: relu",0.911905,0.055830
"Layers: (20,), Activation: tanh",0.933862,0.036724


### Summary of Findings

1. **Performance**:
   - The configuration with **20 hidden neurons and Tanh activation** achieved the highest mean score of **0.933862**. This indicates that it is the most effective at predicting Chronic Kidney Disease among the tested configurations.

2. **Effect of Activation Functions**:
   - Both configurations with **10 hidden neurons**, regardless of whether logistic or ReLU activations were used, produced identical mean scores of **0.911905**. This suggests that adding more complexity (in terms of hidden layers or neurons) can improve model performance, since the single-layer configurations are less effective.

3. **Standard Deviations**:
   - The standard deviation for the Tanh configuration is the lowest at **0.036724**, indicating that this model is providing consistent predictions across different data folds. In contrast, the standard deviations for the logistic and ReLU configurations are higher (**0.055830**), which suggests slightly more variability in performance.